pkg/apis/resource/structured/namedresources/validation/validation.go

```diff
@@ -63,6 +64,8 @@ func validateInstances(instances []resource.NamedResourcesInstance, fldPath *fie
 	return allErrs
 }
 
+var semverRe = regexp.MustCompile(`^(0|[1-9]\d*)\.(0|[1-9]\d*)\.(0|[1-9]\d*)(?:-((?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)(?:\.(?:0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*))*))?(?:\+([0-9a-zA-Z-]+(?:\.[0-9a-zA-Z-]+)*))?$`)
```

Comment:
I like the idea of an in-tree regex defining the format.

That ensures this can remain stable regardless of what happens with the runtime library we parse with.

It also means we can render out the validation into openapi schemas.

I see this is copied straight off the semver website, so I don't doubt it's correct, but a little extraction of the repeated patterns and cleanup of the non-capturing group decorations would make this a lot more understandable:

```go
var (
	numericIdentifier = `(0|[1-9]\d*)`

	preReleaseIdentifier = `(0|[1-9]\d*|\d*[a-zA-Z-][0-9a-zA-Z-]*)`

	buildIdentifier = `[0-9a-zA-Z-]+`

	semverRe = regexp.MustCompile(`^` +

		// dot-separated version segments (e.g. 1.2.3)
		numericIdentifier + `\.` + numericIdentifier + `\.` + numericIdentifier +

		// optional dot-separated prerelease segments (e.g. -alpha.PRERELEASE.1)
		`(-` + preReleaseIdentifier + `(\.` + preReleaseIdentifier + `)*)?` +

		// optional dot-separated build identifier segments (e.g. +build.id.20240305)
		`(\+` + buildIdentifier + `(\.` + buildIdentifier + `)*)?` +

		`$`)
)
```

Reply:
Agreed, and changed.

Reply:
As this simplified the regexp, I also added some more test cases.